File: technique/two_sum/twosum_hashing.rb
def twosum_hashing(nums, target)
  hash = {}
  nums.each_with_index do |num, i|
    hash[num] = i
  end

  nums.each_with_index do |num, i|
    complement = target - num
    if hash.key?(complement) && hash[complement] != i
      return [i, hash[complement]]
    end
  end
  nil
end

p twosum_hashing([2, 7, 11, 15], 13)

File: technique/two_sum/twosum_hashing.py
def twosum_hashing(nums, target):
  nums_map = {}
  for i, num in enumerate(nums):
    complement = target - num
    if complement in nums_map:
      return [nums_map[complement], i]
    nums_map[num] = i
  return [-1, -1]


print(twosum_hashing([2, 7, 11, 15], 9))